cli/flox-rust-sdk/src/providers/services.rs

```diff
@@ -405,4 +581,202 @@ mod tests {
+        _temp_dir: TempDir,
+        socket: PathBuf,
+        child: std::process::Child,
+    }
+
+    impl TestProcessComposeInstance {
+        /// Start a `process-compose` instance with the given [ProcessComposeConfig].
+        /// Wait for the socket to appear before returning.
+        ///
+        /// Panics if the socket doesn't appear after 5 tries with backoff.
+        fn start(config: &ProcessComposeConfig) -> Self {
+            let temp_dir = TempDir::new().unwrap();
+
+            let config_path = temp_dir.path().join("config.yaml");
+            let socket = temp_dir.path().join("S.process-compose");
+            write_process_compose_config(config, &config_path).unwrap();
+
+            let mut cmd = Command::new(&*PROCESS_COMPOSE_BIN);
+
+            // apparently it doesn't do this automatically even though it's not connected to a tty...
+            cmd.env("NO_COLOR", "1");
+            cmd.arg("--unix-socket")
+                .arg(&socket)
+                .arg("--config")
+                .arg(config_path)
+                .arg("--tui=false")
+                .arg("up")
+                .stdout(Stdio::null())
+                .stderr(Stdio::inherit());
+
+            // Dropping the child as stopping is handled via a process-compose command.
+            let child = cmd.spawn().unwrap();
+
+            let max_tries = 5;
+            for backoff in 1..max_tries {
+                println!("waiting for socket to exist");
+                thread::sleep(Duration::from_millis(100 * backoff));
+
+                // For now just check if the socket exists.
+                // Processes _may_ have not started yet, or the socket is unresponsive.
+                // We can't really check if the process is running,
+                // as it may have already exited.
+                // We could chek if the socket can be connected to
+                // or try to read ProcessStates, if the current approach leads to flaking tests.
+                if socket.exists() {
+                    break;
+                }
+
+                if backoff == max_tries {
+                    panic!("socket never appeared");
+                }
+            }
+
+            Self {
+                _temp_dir: temp_dir,
+                socket,
+                child,
+            }
+        }
+
+        /// Get the path to the socket.
+        fn socket(&self) -> &Path {
+            self.socket.as_ref()
+        }
+
+        /// Stop the `process-compose` instance.
+        fn stop(self) {
+            drop(self)
+        }
+    }
+
+    /// Try to stop the process-compose instance by sending a SIGTERM
+    /// to the process-compose process, which will stop all services.
+    impl std::ops::Drop for TestProcessComposeInstance {
+        fn drop(&mut self) {
+            let term_result = nix::sys::signal::kill(
+                nix::unistd::Pid::from_raw(self.child.id() as i32),
+                nix::sys::signal::SIGTERM,
```

Comment:
question nonblocking: why SIGTERM instead of `process-compose down`?

Reply:
Dan and I discussed both and found them to be equivalent
Sigterm was the first I looked into to implement here.
I don't have a strong preference though.

Reply:
I added a comment at least

Reply:
We'd also need to copy the workaround for graceful shutdown until we upgrade `process-compose`, so simpler seems fine for now and we can switch later.